sled-agent/config-reconciler/src/ledger/legacy_configs.rs

```diff
@@ -0,0 +1,417 @@
+use slog_error_chain::InlineErrorChain;
+
+use super::CONFIG_LEDGER_FILENAME;
+
+const LEGACY_DISKS_LEDGER_FILENAME: &str = "omicron-physical-disks.json";
+const LEGACY_DATASETS_LEDGER_FILENAME: &str = "omicron-datasets.json";
+const LEGACY_ZONES_LEDGER_FILENAME: &str = "omicron-zones.json";
+
+pub(super) async fn convert_legacy_ledgers(
+    config_datasets: &[Utf8PathBuf],
+    log: &Logger,
+) -> Option<OmicronSledConfig> {
+    let disk_paths = config_datasets
+        .iter()
+        .map(|p| p.join(LEGACY_DISKS_LEDGER_FILENAME))
+        .collect::<Vec<_>>();
+    let dataset_paths = config_datasets
+        .iter()
+        .map(|p| p.join(LEGACY_DATASETS_LEDGER_FILENAME))
+        .collect::<Vec<_>>();
+    let zone_paths = config_datasets
+        .iter()
+        .map(|p| p.join(LEGACY_ZONES_LEDGER_FILENAME))
+        .collect::<Vec<_>>();
+
+    let loaded_ledgers = futures::join!(
+        Ledger::<OmicronPhysicalDisksConfig>::new(log, disk_paths.clone()),
+        Ledger::<DatasetsConfig>::new(log, dataset_paths.clone()),
+        Ledger::<OmicronZonesConfigLocal>::new(log, zone_paths.clone()),
+    );
+
+    let (disks, datasets, zones) = match loaded_ledgers {
+        // If we have all three, proceed below to our conversion.
+        (Some(disks), Some(datasets), Some(zones)) => {
+            (disks.into_inner(), datasets.into_inner(), zones.into_inner())
+        }
+
+        // If none of them exist, we have no conversion to do.
+        (None, None, None) => return None,
+
+        // Any other combo is terrible: we have one or two legacy ledgers, but
+        // are missing the other two or one. This should only be possible if we
+        // were interrupted mid-rack-setup, which already requires
+        // clean-slate'ing (which should remove these config files!). Panic if
+        // we hit this case: it is not safe to continue with sled-agent startup
+        // if we are partially configured.
+        (disks, datasets, zones) => {
+            error!(
+                log,
+                "Found partial legacy ledgers; unsafe to proceed";
+                "found-legacy-disks" => disks.is_some(),
+                "legacy-disk-paths" => ?disk_paths,
+                "found-legacy-datasets" => datasets.is_some(),
+                "legacy-dataset-paths" => ?dataset_paths,
+                "found-legacy-zones" => zones.is_some(),
+                "legacy-zone-paths" => ?zone_paths,
+            );
+            panic!(
+                "Found partial legacy ledgers; unsafe to proceed (\
+                 found-legacy-disks: {}, \
+                 legacy-disk-paths: {disk_paths:?}, \
+                 found-legacy-datasets: {}, \
+                 legacy-dataset-paths: {dataset_paths:?}, \
+                 found-legacy-zones: {}, \
+                 legacy-zone-paths: {zone_paths:?} \
+                )",
+                disks.is_some(),
+                datasets.is_some(),
+                zones.is_some(),
```

Comment:
Is this really impossible? I thought RSS called three different endpoints on sleds to write out disks, dataset, and zone information.

Wouldn't it be *theoretically* possible to hit this function partway through this operation?

(I don't think this is likely, just trying to grok the choice of panic vs error. It also seems impossible once we switch RSS to not use the legacy format at all)